### PR TITLE
Adding guideline section for events.

### PR DESCRIPTION
**Pull request by JosephMaxwell**
Two rules for events:

1. Treating all event inputs as immutable. If a value is not modified, this allows for Magento's expansion to using queue-based systems.
2. The second one is pretty easy: specific events.

### CODE REVIEW

**Comment by xcomSteveJohnson** [comment]
@buskamuza Please review this PR.

File: guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md [hunk 1]
```diff
@@ -471,6 +471,39 @@ We are reviewing this section and will publish it soon.
...
+             */
+            $transport->setData('output_return_value', true);
+        }
+    }
+}
+```
+{%endhighlight%}
+{% endcollapsible %}
+
+14.2. Use as specific events as possible. Don't use a `global` event when the area impacted is just `frontend`.
```
Comment (JosephMaxwell) [inline guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md]: Yes, that is correct.

File: guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md [hunk 2]
```diff
@@ -471,6 +471,39 @@ We are reviewing this section and will publish it soon.
 
 13.4. Exception in a single CLI command SHOULD NOT break the CLI framework; running other commands SHOULD still be possible.
 
+## 14. Events
+
+14.1. All values (including objects) passed to event MUST NOT be modified in the event observer. Use plugins instead for this application.
```
Comment (dmrachkovskyi) [inline guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md]: >passed to event

Please add article: _passed to **an** event_

>Use plugins instead for this application

May I suggest another wording: _Instead, use plugins for this application.

File: guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md [hunk 3]
```diff
@@ -471,6 +471,39 @@ We are reviewing this section and will publish it soon.
...
+             */
+            $transport->setData('output_return_value', true);
+        }
+    }
+}
+```
+{%endhighlight%}
+{% endcollapsible %}
+
+14.2. Use as specific events as possible. Don't use a `global` subscription to an event when the area impacted is just `frontend`.
```
Comment (dmrachkovskyi) [inline guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md]: >Use as specific events as possible

May I suggest keeping the tone and style of the entire document: _Events used SHOULD be as specific as possible._

> Don't use a `global` subscription ... `frontend`.

Same about tone & style here: _A `global` subscription to an event SHOULD NOT be used when the area impacted is just `frontend`._

Comment (JosephMaxwell) [inline guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.md]: Hello, @dmrachkovskyi, thank for the feedback. I made a few changes: let me know if you have further questions.

**Comment by tanberry** [comment]
Thanks @JosephMaxwell for the contribution, and all for the work to review and merge it! Yay!

**Pull request merged**